src/node.cc

```diff
@@ -2432,33 +2438,83 @@ static void OnMessage(Local<Message> message, Local<Value> error) {
   FatalException(Isolate::GetCurrent(), error, message);
 }
 
+static Maybe<bool> ProcessEmitWarningGeneric(Environment* env,
+                                             const char* warning,
+                                             const char* type = nullptr,
+                                             const char* code = nullptr) {
+  HandleScope handle_scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  Local<Object> process = env->process_object();
+  Local<Value> emit_warning;
+  if (!process->Get(env->context(),
+                    env->emit_warning_string()).ToLocal(&emit_warning)) {
+    return Nothing<bool>();
+  }
+
+  if (!emit_warning->IsFunction()) return Just(false);
+
+  int i = 0;
+  Local<Value> args[3];  // warning, type, code
+
+  // The caller has to be able to handle a failure anyway, so we might as well
+  // do proper error checking for string creation.
+  if (!String::NewFromUtf8(env->isolate(),
+                           warning,
+                           v8::NewStringType::kNormal).ToLocal(&args[i++])) {
+    return Nothing<bool>();
+  }
+  if (type != nullptr) {
+    if (!String::NewFromOneByte(env->isolate(),
+                              reinterpret_cast<const uint8_t*>(type),
+                              v8::NewStringType::kInternalized)
+                                  .ToLocal(&args[i++])) {
+      return Nothing<bool>();
+    }
+    if (code != nullptr &&
+        !String::NewFromOneByte(env->isolate(),
+                                reinterpret_cast<const uint8_t*>(code),
+                                v8::NewStringType::kInternalized)
+                                    .ToLocal(&args[i++])) {
+      return Nothing<bool>();
+    }
+  }
+
+  // MakeCallback() unneeded because emitWarning is internal code, it calls
+  // process.emit('warning', ...), but does so on the nextTick.
+  if (emit_warning.As<Function>()->Call(env->context(),
```

Comment:
If `process.emitWarning` is supposed to be overwritable. Then we should use `MakeCallback`, right?

Reply:
Mh … I am not sure. There is usually always a JS call stack below, right?

Reply:
Well, that won't be true for the `domain` `MakeCallback` deprecation. Anything could call `MakeCallback` in that case.

Reply:
Hm. True. But that would put us in a pretty difficult position if `process.domain` happens to be set in that case (by accident), since `process` would be the resource object?

So, I am okay with changing this but I’m not sure where we’d get the async context to use for `MakeCallback`?

Reply:
> So, I am okay with changing this but I’m not sure where we’d get the async context to use for MakeCallback?

Yes. This is a general problem. We have `process.on('beforeExit')`,  `process.on('exit')` without context as well. In this case I think there are two options:
* Have a general `async_context` for the `process`, and use that.
* Create a new `async_context` with `async_context_.async_id` as the `triggerAsyncId`.

The latter option is better for this case but I don't see it working in general for all `ProcessEmitWarningGeneric` use cases.